ouroboros-consensus/src/Ouroboros/Consensus/Storage/LedgerDB/HD/DiffSeq.hs

```diff
@@ -0,0 +1,354 @@
+{-# LANGUAGE ConstraintKinds            #-}
+{-# LANGUAGE DataKinds                  #-}
+{-# LANGUAGE DeriveAnyClass             #-}
+{-# LANGUAGE DeriveFunctor              #-}
+{-# LANGUAGE DeriveGeneric              #-}
+{-# LANGUAGE DerivingVia                #-}
+{-# LANGUAGE FlexibleContexts           #-}
+{-# LANGUAGE FlexibleInstances          #-}
+{-# LANGUAGE GeneralisedNewtypeDeriving #-}
+{-# LANGUAGE MultiParamTypeClasses      #-}
+{-# LANGUAGE ScopedTypeVariables        #-}
+{-# LANGUAGE TypeFamilies               #-}
+
+{- | Sequences of diffs for ledger tables.
+
+   These diff sequences are an instantiation of a strict finger tree with root
+   measures. The tree/sequence itself contains diffs and slot information, while
+   the root measure is the total sum of all diffs in the sequence. The internal
+   measure is used to keep track of sequence length and maximum slot numbers.
+
+   The diff datatype that we use forms a @'Group'@, which allows for relatively
+   efficient splitting of finger trees with respect to recomputing measures by
+   means of the @'invert'@ operation that the @'Group'@ type class requires.
+   Namely, if either the left or right part of the split is small in comparison
+   with the input sequence, then we can subtract the diffs in the smaller part
+   from the root measure of the input to (quickly) compute the root measure of
+   the /other/ part of the split. This is much faster than computing the root
+   measures from scratch by doing a linear-time pass over the elements of the
+   split parts, or a logarithmic-time pass over intermediate sums of diffs in
+   case we store cumulative diffs in the nodes of the finger tree.
+
+   === Example of fast splits
+
+   As an analogy, consider this example: we have a sequence of consecutive
+   integer numbers @xs = [1..n]@ where @n@ is large, and we define the root
+   measure of the sequence to be the total sum of these numbers, @rmxs = sum
+   [1..n]@ (we assume @rmxs@ is fully evaluated). Say we split this sequence of
+   integer numbers at the index @2@, then we get /left/ and /right/ parts of the
+   split @ys@ and @zs respectively.
+
+   > splitAt 2 xs = (ys, zs) = ([1..2], [3..n])
+
+   How should we compute we the root measure @rmys@ of @ys@? Since @ys@ is
+   small, we can just compute @rmys = sum [1..2]@. How should we compute the
+   root measure @rmzs@ of @zs@? We should not compute @rmzs = sum [3..n]@ in this
+   case, since @n@ is large. Instead, we compute @rmzs = rmxs - rmys@, which
+   evaluates to its result in time that is linear in the length of @ys@, in this
+   case @O(2)@.
```

Comment:
`O(1)` actually, since the constant is irrelevant, isn't it?

Reply:
Done